src/integer_mod_q/mat_zq/solve.rs

```diff
@@ -58,42 +62,283 @@ impl MatZq {
+                break;
+            }
+
+            // Try to solve the system under the current modulus.
+            if let Some((row_nr, inv)) =
+                find_invertible_entry_column(&matrix_identity_base_gauss, column_nr, &used_rows)
+            {
+                matrix_identity_base_gauss.gauss_row_reduction(row_nr, column_nr, inv);
+
+                if row_count != row_nr {
+                    matrix_identity_base_gauss
+                        .swap_rows(row_count, row_nr)
+                        .unwrap();
+
+                    permutation.swap(row_count.try_into().unwrap(), row_nr.try_into().unwrap());
+                }
+
+                // Save the position of `1` for that row.
+                indices.push((row_count, column_nr));
+                used_rows.push(indices[indices.len() - 1].0);
+                row_count += 1;
+            // Search for a not invertible entry to divide the modulus.
+            } else if let Some(row_nr) =
+                find_not_invertible_entry_column(&matrix_identity_base_gauss, column_nr, &used_rows)
+            {
+                // Factorize the modulus with the found entry, compute solutions
+                // for the system under the split modulus and use the
+                // Chinese Remainder Theorem to compute a solution based on these
+                // sub-solutions.
+                let entry: Z = matrix_identity_base_gauss
+                    .get_entry(row_nr, column_nr)
+                    .unwrap();
+                self.factorization_and_crt(y, entry);
+            }
         }
 
-        // check if we have computed a solution and return `None`
-        if &(self * &out) != y {
+        // Return `None` if the matrix has no full rank.
+        if indices.is_empty()
+            || indices[indices.len() - 1].0 + 1 < matrix_identity_base_gauss.get_num_rows()
+        {
             return None;
         }
 
+        // Pick the first columns out of the original matrix that form an invertible matrix.
+        // Those columns exist, since the matrix was checked to be full rank.
+        let mut invertible_matrix = MatZq::new(
+            matrix_identity_base_gauss.get_num_rows(),
+            matrix_identity_base_gauss.get_num_rows(),
+            self.get_mod(),
+        );
+        for (current_column, (_row_nr, column_nr)) in indices.iter().enumerate() {
+            invertible_matrix
+                .set_column(current_column, self, *column_nr)
+                .unwrap();
+        }
+
+        // The inverse of the previously picked square matrix consists of the last
+        // columns of `matrix_identity_base_gauss`, since we concatenated an identity matrix.
+        let matrix_identity_gauss = matrix_identity_base_gauss.change_modulus(&self.get_mod());
+        let mut matrix_base_inv = MatZq::new(
+            matrix_identity_gauss.get_num_rows(),
+            matrix_identity_gauss.get_num_rows(),
+            &self.get_mod(),
```

Comment:
sometimes you use &self.get_mod() and sometimes only self.get_mod(), only do the latter

Reply:
It does not work in MatZq::from

Reply:
included it in the refactoring